clustered/server/src/test/java/org/ehcache/clustered/server/offheap/ChainMapExtensionTest.java

```diff
@@ -0,0 +1,378 @@
+      } else {
+        assertThat(map.get("foo" + i), emptyIterable());
+      }
+    }
+  }
+
+  private OffHeapChainMap<String> getChainMapWithExtendedStorageEngine() {
+    PageSource chainSource = new UnlimitedPageSource(new OffHeapBufferSource());
+    PageSource extendedSource = new UnlimitedPageSource(new OffHeapBufferSource());
+    Factory<? extends ChainStorageEngine<String>> factory = ExtendedOffHeapChainStorageEngine.createFactory(chainSource,
+      StringPortability.INSTANCE, 4096, 4096, false, false, extendedSource);
+    return new OffHeapChainMap<>(chainSource, factory);
+  }
+
+  private OffHeapChainMap<String> getNewMap(ExtendedOffHeapChainStorageEngine<String> ese) {
+    PageSource chainSource = new UnlimitedPageSource(new OffHeapBufferSource());
+    Factory<? extends ChainStorageEngine<String>> factory = OffHeapChainStorageEngine.createFactory(chainSource,
+      StringPortability.INSTANCE, 4096, 4096, false, false);
+    OffHeapChainStorageEngine<String> storageEngine = (OffHeapChainStorageEngine<String>) factory.newInstance();
+    ReadWriteLockedOffHeapClockCache<String, InternalChain> newMap =
+      new ReadWriteLockedOffHeapClockCache<>(chainSource, storageEngine);
+    ese.replayIntoMap(newMap);
+    return new OffHeapChainMap<>(newMap, storageEngine);
+  }
+
+  private static ByteBuffer buffer(int i) {
+    ByteBuffer buffer = ByteBuffer.allocate(i);
+    while (buffer.hasRemaining()) {
+      buffer.put((byte) i);
+    }
+    return (ByteBuffer) buffer.flip();
+  }
+
+  private static Matcher<Element> element(final int i) {
+    return new TypeSafeMatcher<Element>() {
+      @Override
+      protected boolean matchesSafely(Element item) {
+        return item.getPayload().equals(buffer(i));
+      }
+
+      @Override
+      public void describeTo(Description description) {
+        description.appendText("element containing buffer[" + i +"]");
+      }
+    };
+  }
+
+  private static final class ExtendedHeaderForTest {
+    private final ByteBuffer data;
+    private final WriteContext writeContext;
+
+    ExtendedHeaderForTest(ByteBuffer buffer, WriteContext writeContext) {
+      this.data = buffer;
+      this.writeContext = writeContext;
+    }
+
+    long getAddress() {
+      return getLong(ADDRESS_OFFSET);
+    }
+
+    void setAddress(long val) {
+      writeContext.setLong(ADDRESS_OFFSET, val);
+    }
+
+    int getHash() {
+      long hashAndSize = getLong(HASH_OFFSET) >> 32;
+      return (int) hashAndSize;
+    }
+
+    int getSize() {
+      long hashAndSize = getLong(HASH_OFFSET);
+      return (int) hashAndSize;
+    }
+
+    void setHashAndSize(int hash, int size) {
+      long val = ((long) hash << 32) | size;
+      writeContext.setLong(HASH_OFFSET, val);
+    }
+
+    private long getLong(int address) {
+      return data.getLong(address);
+    }
+  }
+
+  public static class ExtendedOffHeapChainStorageEngine<K> extends OffHeapChainStorageEngine<K> {
```

Comment:
This class looks a bit too much for a test. Why is it that complicated?

Reply:
hmm.. this is only way to test the extension capabilities I added..by adding a storage engine that extends the offheap chain storage engine and uses the extended header to keep track of additional stuff.

One thing I could do is to have the extended storage engine to be a separate class in the test package.. So that the test looks smaller and concise. wdyt?

Reply:
No. I won't change anything. It is the complexity of `ExtendedOffHeapChainStorageEngine` that I don't like.